### DISCUSSION TITLE
Usage of Utility Module in Standalone Mode of mife

### DISCUSSION THREAD

**Discussion by PDGallus** [WJSoftware/vite-plugin-single-spa#116] [Q&A] [answered]
Hello everyone.
As Jose suggested, I am sharing my problem again here in the discussion area.

**My Problem:**

- Standalone mode cannot resolve the import of util-module, due to the absence of ImportMaps
- Manual addition of the ImportMaps is also not possible due to vite peculiarity

**Goal**

- Make mife with import of util-module work in standalone mode

**Context**
The utility module has logic for an event bus and state management that is to be potentially distributed in all mife. Both the event bus and the state management are framework agnostic.

What I am trying to achieve is that I can start a mife in standalone mode and then use a utility module in the mife running independently.

In the standard mode of the Single-SPA, I can also obtain the module as normal via ImportMaps, but not in standalone mode.

One circumstance that makes the whole thing even more difficult is that all applications are in their own repos, so there is no mono repo and it is also not an option to throw everything into a mono repo.

Jose suggested writing a small plugin that provides a fake module or copies the utility module. However, I could imagine that copying would not work because the code base is not in a mono repo. Unless I have misunderstood something.
But I need the real function of my utility module, because the applications have to react to state or listen for events in order to do something and behave correctly. So if I want to develop such an application locally, it would of course be best if I could simply load the module.

Jose, could you perhaps explain in a little more detail what your two approaches would be. I have to admit that I have never written a plugin for vite before, so I have no idea what exactly I can achieve with it, or what is possible.

Or maybe the possible options have changed since I shared the info about the missing possibility of a mono repo?

**Comment by webJose** [comment, marked as answer]
Hello, @PDGallus.  Indeed, what you describe is an "issue" I also have at work.  In my case, and at least for now, I only need a configuration module that is only available from the root MFE.  It is a super, mega simple export of a JS object, where the main property is the environment (possible values are `Development`, `Test`, `PreProduction` and `Production`).

So what do I have and what did I do?  Let's see.

## What I have

All non-root MFE's require this configuration file, so the appropriate configuration for each environment is used.  I use the `wj-config` NPM package (also of my crafting) for configuration, and every non-root MFE has a `config.js` file similar to this:

```javascript
import wj-config from "wj-config";
import mainConfig from "@appDomain/client-env"; // <------ This is an alias registered in the root MFE's import map!

... // etc.
```

## Possible Scenarios

There are 3 scenarios to cover:

1. **Deployed build**:  This is what we deploy to Kubernetes, and is the resultant of `npm run build`.
2. **Development in single-spa mode**:  We usually work this way:  We open the application in a deployed environment (say, `Test`), and use `import-map-overrides` to mount the one MFE we are working on in there.  We run that MFE using `npm run dev`, so no actual bundling is taking place.  This is the recommended scenario when working with `single-spa` as stated in their documentation.
3. **Standalone mode**:  This is where we don't mount the MFE in a deployed environment and merely run that one MFE we will be working on, using `npm run dev`.

### Deployed Build

This is the easiest.  In order to successfully bundle the project,  we simply mark the `@appDomain/client-env` module as external in the rollup options:

```javascript
// vite.config.js

export default defineConfig({
    build: {
        rollupOptions: {
            external: ['@appDomain/client-env']
        }
    }
    ... // etc.
});
```

What this does is tell rollup not to attempt to obtain that module's source code and merely leave the `import` instruction untouched.

### Development in single-spa Mode

This is everyone's favorite, and while it looks more like the deployed build, is actually almost identical to standalone mode.

Even when running under a page provided by a root MFE, Vite, while in serve mode (`npm run dev`) will try to resolve the module id `@appDomain/client-env`.  The `external` setting has no effect because that's for bundling (rollup is all about bundling) and Vite in serve mode doesn't bundle.

### Standalone Mode

This scenario, for the purposes of your problem, is identical to the previous one.

---

So what to do?  Well, since Vite won't desist in trying to resolve the external modules, we are simply bound to provide them.  How?  With a very simple plug-in.  Vite/rollup plug-ins are mere functions that return an object with functions, so you can write the plug-in even inside `vite.config.js` if you want.

## The Plug-In

Plug-ins are usually written in the form of factory functions, so let's follow the pattern:

```javascript
// Export this function if you wrote it outside of vite.config.js.

function devPlugin() {
    return {
        name: 'devPlugin',
        resolveId(source) {
            if (source === '@appDomain/client-env') {
                return source;
            }
            return null;
        },
        load(id) {
            if (id === '@appDomain/client-env') {
                return 'export default { environment: "Development" };';
            }
        }
    };
}
```

In all honesty, I think we can eliminate the `resolveId()` function and just have the plug-in with a `load()` function.  The `load()` function's purpose is to provide the module's source code.  this is why you see I return valid javascript module in the form of a string.  This is of course, a simplified example where I simply export an object with the one `environment` property that I truly need.  This is a "fake", or "reduced" version of what the root MFE offers.  I go this route because it satisfies my current requirements.

Now, you say you need the real thing, not a fake.  There is no choice:  As seen in the sample above, you **must** have access to the source code so the `load()` function can return it.  Arrange your repos somehow or attempt any other magic you want.  There is no escape:  You have to have the source code for the real thing.

What would I do?  Well, Intel has several private NPM repositories.  If I ever were in your position, I would set up a private NPM repository and create a package out of the utility module the root provides.  This would mean I'd probably have a repository for the utility module, and would use that repository to produce the NPM package.  This means that the root MFE would no longer contain the module's source code, and would instead use the NPM package as well.

Going back to the plug-in solution with the NPM package in mind, you should at this point realize:  I don't need the plug-in solution at all!  You would just install the NPM package for the utility module just like you did in in the root MFE and problem solved, right?  The answer is **maybe**.  It would be yes if the NPM package's name were `@appDomain/client-env` (to continue with my example).  But what if it were named, say `app-client-env`?  Then the import statement, to work, would need to be `import mainConfig from "app-client-env";` instead of `... from "@appDomain/client-env"`.  The plug-in is still needed.

If you found the proposed solution of a private NPM package appealing, then simplify the plug-in above, not by removing `resolveId()`, but by removing `load()`, and then make `resolveId()` return the true NPM package name `app-client-env`.  Then Vite's own implementation of `load()` will search the `node_modules/` folder, will find the source code, and all will be well for the `npm run dev` scenarios.

This would be it:

```javascript
// Export this function if you wrote it outside of vite.config.js.

function devPlugin() {
    return {
        name: 'devPlugin',
        resolveId(source) {
            if (source === '@appDomain/client-env') {
                return 'app-client-env';
            }
            return null;
        }
    };
}
```

**Reply by PDGallus** [reply]
@webJose you're awesome. Thank you very much for your time and effort. I will work through it and come back here and report how it goes. 
I won't be able to do it next week because I'm on vacation. But I'll get back to you after that.

Have yourself a great time.

**Reply by webJose** [reply]
BTW, I just remembered that you can do aliasing without plug-ins.  See [this piece of document](https://vitejs.dev/config/shared-options.html#resolve-alias).  Maybe you can use it and not need the plug-in.  It says, though that one must return full file paths.  Unsure if that applies to NPM packages.  There's an explicit mention of an `npm:` prefix I don't know about.  Anyway, might be worth the try.

**Reply by PDGallus** [reply]
Hey @webJose!
After my vacation, I worked through your awesome answer. Thanks again for the effort and work you put into it. Due to the fact that the application is in distributed teams, I will go with the suggested way, on the part of single-SPA, and use the import map overrides. This way, everyone can hook into the existing system and test their code. This seems to me to be the easiest way for everyone at the moment.

Cheers
